File: models/job.go
package models

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
	"time"
)

// Is there better way to support enums?
const (
	StatusInitial    = "initial"
	StatusProcessing = "processing"
	StatusFailed     = "failed"
	StatusCompleted  = "completed"
)

// Job is Job
// I've found this interesting: https://golang.org/pkg/database/sql/#Rows.Scan
//
//go:generate reform
//reform:jobs
type Job struct {
	ID          int32     `reform:"id,pk" json:"id"`
	Type        string    `reform:"type" json:"type"`
	Path        string    `reform:"path" json:"path"`
	Sid         *string   `reform:"sid" json:"sid"`
	Input       *string   `reform:"input" json:"-"`
	Output      *string   `reform:"output" json:"output"`
	Status      string    `reform:"status" json:"status"`
	TreeStatus  *string   `json:"treeStatus"`
	RetriesLeft int32     `reform:"retries_left" json:"retriesLeft"`
	CreatedAt   time.Time `reform:"created_at" json:"createdAt"`
}

// GetInput is getter for Input
func (job *Job) GetInput() JSONDictionary {
	var dict JSONDictionary
	_ = json.Unmarshal([]byte(*job.Input), &dict)
	return dict
}

// SetInput is setter for Input
func (job *Job) SetInput(dict JSONDictionary) string {
	valueAsBytes, _ := json.Marshal(dict)
	value := string(valueAsBytes)
	job.Input = &value
	return value
}

// GetOutput is getter for Output
func (job *Job) GetOutput() JSONDictionary {
	dict := JSONDictionary{}
	if job.Output == nil {
		return dict
	}
	_ = json.Unmarshal([]byte(*job.Output), &dict)
	return dict
}

// SetOutput is setter for Output
func (job *Job) SetOutput(dict JSONDictionary) string {
	valueAsBytes, _ := json.Marshal(dict)
	value := string(valueAsBytes)
	job.Input = &value
	return value
}

// GetRootID returns root ID for this job's tree
func (job *Job) GetRootID() (int32, bool) {
	if job.Path == "" {
		return 0, false
	}
	ids := strings.Split(job.Path, "/")
	rootID, _ := strconv.Atoi(ids[0])
	return int32(rootID), true
}

// SetParent initialize job's path by it's parent
func (job *Job) SetParent(parent *Job) {
	if parent == nil {
		return
	}
	parentIDString := strconv.Itoa(int(parent.ID))
	if parent.Path == "" {
		job.Path = parentIDString
	} else {
		job.Path = strings.Join([]string{parent.Path, parentIDString}, "/")
	}
}

// IsRoot returns whether the job is root or not
func (job *Job) IsRoot() bool {
	_, hasRoot := job.GetRootID()
	return !hasRoot
}

// IsCompleted returns if job is completed
func (job *Job) IsCompleted() bool {
	return job.Status == StatusCompleted || job.Status == StatusFailed
}

// IsFailed returns if job is failed
func (job *Job) IsFailed() bool {
	return job.Status == StatusFailed
}

// BuildJob is a builder helper
func BuildJob(jobType string, input JSONDictionary, parent *Job) *Job {
	job := &Job{
		Type: jobType,
	}
	job.SetInput(input)
	job.SetParent(parent)
	return job
}

func (job *Job) IsParentOf(childJob *Job) bool {
	pathWithParentID := strings.Trim(fmt.Sprintf("%v/%v", job.Path, job.ID), "/")
	return strings.Contains(childJob.Path, pathWithParentID)
}

func (job *Job) IsChildOf(parentJob *Job) bool {
	pathWithParentID := strings.Trim(fmt.Sprintf("%v/%v", parentJob.Path, parentJob.ID), "/")
	return strings.Contains(job.Path, pathWithParentID)
}

File: gui/src/models/job.ts
import TreeLeaf from './tree_leaf'

export type Status = 'initial' | 'processing' | 'failed' | 'completed'

type Props = {
  id: number
  type: string
  path: string
  sid: string
  input: string
  output: string
  status: Status
  createdAt: string

  childs: Job[]
}

class Job implements Props, TreeLeaf {
  id: number = 0
  type: string = ''
  path: string = ''
  sid: string = ''
  input: string = ''
  output: string = ''
  status: Status = 'initial'
  createdAt: string = ''

  childs: Job[] = []

  constructor(props: Partial<Props>) {
    if(props.id) { this.id = props.id }
    if(props.type) { this.type = props.type }
    if(props.path) { this.path = props.path }
    if(props.sid) { this.sid = props.sid }
    if(props.input) { this.input = props.input }
    if(props.output) { this.output = props.output }
    if(props.status) { this.status = props.status }
    if(props.createdAt) { this.createdAt = props.createdAt }

    if(props.childs) { this.childs = props.childs }
  }

  isRoot() {
    return this.path === ''
  }

  parentID() {
    if (this.isRoot()) {
      return null
    }
    const ids = this.path.split('/').map(i => parseInt(i))
    return ids[ids.length - 1]
  }

  digest() {
    return `${this.id}${this.status}`
  }

  static deserialize(json: any): Job {
    const job = new Job(json as Partial<Props>)
    return job
  }
}

export default Job